### PR TITLE
Scabbard CLI

### PR DESCRIPTION
**Pull request by ltseeley**
To test:
- Run the Grid on Splinter example and setup a circuit
- Download the xo smart contract (https://build.sawtooth.me/job/Sawtooth-Hyperledger/job/sawtooth-sdk-rust/job/master/lastSuccessfulBuild/artifact/build/scar/xo_0.4.1.scar)
- Add the xo smart contract's directory as a volume for the scabbard-cli service in docker-compose-installed.yaml (mapped to `/usr/share/scar`)
- Run the scabbard-cli image using `docker-compose -f docker-compose-installed.yaml run scabbard-cli bash`
- In the scabbard-cli container:
  - `mkdir -p /root/.splinter/keys`
  - `echo "<alpha-gridd-pub-key>" > /root/.splinter/keys/gridd.pub`
  - `echo "<alpha-gridd-priv-key>" > /root/.splinter/keys/gridd.priv`
  - `scabbard cr create sawtooth_xo --owner $(cat /root/.splinter/keys/gridd.pub) -k gridd -U 'http://splinterd-alpha:8085' --service-id 'my-grid-circuit::grid-scabbard-a'`
  - `scabbard upload /usr/share/scar/xo_0.4.1.scar -k gridd -U 'http://splinterd-alpha:8085' --service-id 'my-grid-circuit::grid-scabbard-a'`
  - `scabbard ns create 5b7349 --owner $(cat /root/.splinter/keys/gridd.pub) -k gridd -U 'http://splinterd-alpha:8085' --service-id 'my-grid-circuit::grid-scabbard-a'`
  - `scabbard perm 5b7349 sawtooth_xo -r -w -k gridd -U 'http://splinterd-alpha:8085' --service-id 'my-grid-circuit::grid-scabbard-a'`
  - `scabbard perm cad11d sawtooth_xo -r -k gridd -U 'http://splinterd-alpha:8085' --service-id 'my-grid-circuit::grid-scabbard-a'`

There is currently no way to verify the success of these actions besides the proposals being committed (seen in the logs).

### CODE REVIEW

File: services/scabbard/src/cli/action/namespace.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        .subcommand(SubCommand::with_name("create").args(&generate_create_or_update_args()))
+        .subcommand(SubCommand::with_name("update").args(&generate_create_or_update_args()))
+        .subcommand(SubCommand::with_name("delete").arg(generate_namespace_arg()))
+}
+
+fn generate_create_or_update_args<'a, 'b>() -> Vec<Arg<'a, 'b>> {
+    vec![
+        generate_namespace_arg(),
+        Arg::with_name("owner")
+            .help("owner of this namespace")
```
Comment (vaporos) [inline services/scabbard/src/cli/action/namespace.rs]: Help text should start with a capital letter to match --help/--version usage.

File: services/scabbard/src/cli/action/permission.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+use clap::{App, Arg, ArgMatches, SubCommand};
+use sabre_sdk::protocol::payload::{
+    Action, CreateNamespaceRegistryPermissionActionBuilder,
+    DeleteNamespaceRegistryPermissionActionBuilder, SabrePayload, SabrePayloadBuilder,
+};
+
+use super::CliError;
+
+pub fn subcommand<'a, 'b>() -> App<'a, 'b> {
```
[the remark this answers is not in the document]
Comment (ltseeley) [reply services/scabbard/src/cli/action/permission.rs]: Removed all single-statement wrapper functions

File: services/scabbard/src/cli/action/smart_permission.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+pub fn subcommand<'a, 'b>() -> App<'a, 'b> {
+    SubCommand::with_name("sp")
+        .about("create, update or delete smart permissions")
+        .subcommand(SubCommand::with_name("create").args(&generate_create_or_update_args()))
+        .subcommand(SubCommand::with_name("update").args(&generate_create_or_update_args()))
+        .subcommand(SubCommand::with_name("delete").args(&generate_org_id_and_name_args()))
+}
+
+fn generate_create_or_update_args<'a, 'b>() -> Vec<Arg<'a, 'b>> {
```
[the remark this answers is not in the document]
Comment (ltseeley) [reply services/scabbard/src/cli/action/smart_permission.rs]: Inlined

File: services/scabbard/src/cli/action/namespace.rs (hunk 1 above)
Comment (ltseeley) [reply services/scabbard/src/cli/action/namespace.rs]: Updated all `about`s and `help`s to start with capital letters

File: services/scabbard/src/cli/key.rs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+fn determine_key_file_path(key: &str) -> Result<PathBuf, CliError> {
+    if key.contains('/') {
+        Ok(PathBuf::from(key))
+    } else {
+        let mut path = dirs::home_dir().ok_or_else(|| {
+            CliError::action_error("failed to load signing key: unable to determine home directory")
+        })?;
+        path.push(".scabbard");
```
Comment (vaporos) [inline services/scabbard/src/cli/key.rs]: We should make this use .splinter (keys for the project, as opposed to keys for a particular CLI).

File: services/scabbard/src/cli/main.rs [hunk 5]
```diff
@@ -0,0 +1,642 @@
...
+                        .conflicts_with("delete"),
+                    Arg::with_name("delete")
+                        .help("Remove all permissions")
+                        .short("d")
+                        .long("delete"),
+                ]),
+            SubCommand::with_name("cr")
+                .about("Create, update, or delete a Sabre contract registry")
+                .subcommand(
+                    SubCommand::with_name("create").args(&[
```
Comment (vaporos) [inline services/scabbard/src/cli/main.rs]: Not all subcommands under cr, ns, and sp have usage help (as seen when doing 'cargo run -- cr --help'.).

File: services/scabbard/src/cli/main.rs [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+        .value_of("service-id")
+        .ok_or_else(|| CliError::MissingArgument("service-id".into()))?;
+    let (circuit_id, service_id) = split_full_service_id(full_service_id)?;
+
+    let wait = matches
+        .value_of("wait")
+        .expect("default not set for --wait")
+        .parse::<u64>()
+        .map_err(|_| {
+            CliError::InvalidArgument(format!("'{}' argument must be a valid integer", "wait"))
```
Comment (vaporos) [inline services/scabbard/src/cli/main.rs]: No need for format!() here.

File: services/scabbard/src/cli/transaction.rs [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+/// # Arguments
+///
+/// * `payload` - a fully populated identity payload
+/// * `signer` - the signer to be used to sign the transaction
+///
+/// # Errors
+///
+/// An error is returned if it occurs during serialization of the provided payload or internally
+/// created `TransactionHeader`, or if a signing error occurs.
+pub fn create_transaction(
```
Comment (vaporos) [inline services/scabbard/src/cli/transaction.rs]: We should really put into_transaction_builder() into the Sabre SDK (would essentially be this code minus the signer).

File: services/scabbard/src/cli/transaction.rs [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+
+use protobuf::{self, Message};
+use sabre_sdk::{
+    protocol::{
+        payload::{Action, SabrePayload},
+        ADMINISTRATORS_SETTING_ADDRESS,
+    },
+    protos::IntoBytes,
+};
+use sawtooth_sdk::messages::{
```
Comment (vaporos) [inline services/scabbard/src/cli/transaction.rs]: We can probably use Transact here and the better builder pattern.

File: services/scabbard/src/cli/transaction.rs [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+
+    Ok(batch)
+}
+
+/// Returns a BatchList containing the provided Batch
+///
+/// # Arguments
+///
+/// * `batch` - a Batch
+pub fn create_batch_list_from_one(batch: Batch) -> BatchList {
```
Comment (vaporos) [inline services/scabbard/src/cli/transaction.rs]: Should be transaction.into_batch_builder() in Transact.

File: services/scabbard/src/cli/error.rs [hunk 10]
```diff
@@ -0,0 +1,137 @@
+// Copyright 2020 Cargill Incorporated
```
Comment (agunde406) [inline services/scabbard/src/cli/error.rs]: Needs to be // Copyright 2018-2020 Cargill Incorporated
(same for rest of the new files)

File: services/scabbard/src/cli/transaction.rs [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+    let mut batch_list = BatchList::new();
+    batch_list.set_batches(protobuf::RepeatedField::from_vec(vec![batch]));
+    batch_list
+}
+
+/// Creates a nonce appropriate for a TransactionHeader
+fn create_nonce() -> String {
```
Comment (vaporos) [inline services/scabbard/src/cli/transaction.rs]: nonce should be random. Transact's builder does this automatically; if we can't use Transact for some reason, we should use it's approach. (protocol/transaction.rs)

File: services/scabbard/src/cli/transaction.rs [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+    let elapsed = Instant::now().elapsed();
+    format!("{}{}", elapsed.as_secs(), elapsed.subsec_nanos())
+}
+
+/// Returns a state address for a given namespace registry
+///
+/// # Arguments
+///
+/// * `namespace` - the address prefix for this namespace
+fn compute_namespace_registry_address(namespace: &str) -> Result<String, CliError> {
```
Comment (vaporos) [inline services/scabbard/src/cli/transaction.rs]: I feel like this should be part of the builder sequence (same for similar fns).

File: services/scabbard/src/cli/main.rs [hunk 13]
```diff
@@ -0,0 +1,642 @@
...
+        .setting(AppSettings::SubcommandRequiredElseHelp)
+        .arg(
+            Arg::with_name("verbose")
+                .help("Log verbosely")
+                .short("v")
+                .multiple(true),
+        )
+        .arg(
+            Arg::with_name("key")
+                .help(
```
Comment (agunde406) [inline services/scabbard/src/cli/main.rs]: long help may provide better formatting

File: services/scabbard/src/cli/error.rs (hunk 10 above)
Comment (ltseeley) [reply services/scabbard/src/cli/error.rs]: Updated

File: services/scabbard/src/cli/key.rs (hunk 4 above)
Comment (ltseeley) [reply services/scabbard/src/cli/key.rs]: Updated

File: services/scabbard/src/cli/main.rs (hunk 5 above)
Comment (ltseeley) [reply services/scabbard/src/cli/main.rs]: Added `about` for all subcommands.

File: services/scabbard/src/cli/main.rs (hunk 6 above)
Comment (ltseeley) [reply services/scabbard/src/cli/main.rs]: Fixed

File: services/scabbard/src/cli/transaction.rs (hunk 11 above)
Comment (ltseeley) [reply services/scabbard/src/cli/transaction.rs]: Updated to follow same approach as transact (using `rand`)

File: services/scabbard/src/cli/main.rs (hunk 13 above)
Comment (ltseeley) [reply services/scabbard/src/cli/main.rs]: Changed multi-line helps to use `long_help`

File: services/scabbard/src/cli/transaction.rs (hunk 7 above)
Comment (ltseeley) [reply services/scabbard/src/cli/transaction.rs]: I agree with this and all the other comments about using Transact/Sabre SDK to do all these things for us. In the interest of getting this merged and the functionality available, would it be fine to open stories to update Sabre and Transact to support the desired behavior and make the corresponding changes here?

Comment (ltseeley) [reply services/scabbard/src/cli/transaction.rs]: Created a story to implement this capability

File: services/scabbard/src/cli/transaction.rs (hunk 8 above)
Comment (ltseeley) [reply services/scabbard/src/cli/transaction.rs]: Created stories to implement all necessary capabilities for Transact and update this to use the Transact builder pattern

File: services/scabbard/src/cli/transaction.rs (hunk 9 above)
Comment (ltseeley) [reply services/scabbard/src/cli/transaction.rs]: Created a story to update this to use the Transact builder pattern

File: services/scabbard/src/cli/transaction.rs (hunk 12 above)
Comment (ltseeley) [reply services/scabbard/src/cli/transaction.rs]: Created a story to implement this capability

File: services/scabbard/src/cli/main.rs [hunk 14]
```diff
@@ -0,0 +1,658 @@
...
+    }
+}
+
+fn run() -> Result<(), CliError> {
+    let app = App::new("scabbard")
+        .version(env!("CARGO_PKG_VERSION"))
+        .author("Cargill")
+        .about("Command line for scabbard")
+        .setting(AppSettings::SubcommandRequiredElseHelp)
+        .arg(
```
Comment (agunde406) [inline services/scabbard/src/cli/main.rs]: Can you make these top level argument flags global so they can be placed anywhere in the line

Comment (ltseeley) [reply services/scabbard/src/cli/main.rs]: Updated all of the optional top-level arguments to be global (required arguments cannot be made global).

File: services/scabbard/src/cli/main.rs [hunk 15]
```diff
@@ -0,0 +1,661 @@
...
+}
+
+fn setup_logging(log_level: log::LevelFilter) {
+    let mut log_spec_builder = LogSpecBuilder::new();
+    log_spec_builder.default(log_level);
+
+    Logger::with(log_spec_builder.build())
+        .format(log_format)
+        .start()
+        .expect("Failed to create logger");
```
Comment (vaporos) [inline services/scabbard/src/cli/main.rs]: this should propogate up as a CliError

Comment (ltseeley) [reply services/scabbard/src/cli/main.rs]: Fixed

**Comment by ltseeley** [comment]
Updated to move all global args (except `verbosity`) into subcommands and put all top-level subcommands behind individual feature flags as per @vaporos's request

File: services/scabbard/src/cli/key.rs [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+    use std::path::Path;
+
+    use sawtooth_sdk::signing::PrivateKey;
+    use serial_test::serial;
+    use tempfile::{tempdir, NamedTempFile};
+
+    const MOCK_PRIV_KEY_HEX: &str =
+        "d31e395bed0d9b2277b25d57523063d7d6b9db802d80549bc1362875cdcb83c6";
+
+    /// Verify that an error is retruned when the key file is not found.
```
Comment (agunde406) [inline services/scabbard/src/cli/key.rs]: retruned -> returned (repeated in the tests below)

Comment (ltseeley) [reply services/scabbard/src/cli/key.rs]: Fixed

**Pull request merged**